language/tools/test-generation/src/abstract_state.rs

```diff
@@ -85,6 +87,102 @@ impl AbstractValue {
+/// may not yet exist in the underlying module. Instead of mutating the underlying module (which
+/// would require an into_inner followd by a freeze) in order to record these instantiations in the
+/// locals signature table, we instead build wrapper around the underlying module containing the
+/// type instantiations, and at the end materialize this updated signature pool into a module. We
+/// also need the ability to quickly determine if an instantiation has already been created, and if
+/// so, at which index. So this also keeps a reverse lookup table of instantiation to
+/// LocalsSignatureIndex.
+#[derive(Debug, Clone)]
+pub struct InstantiableModule {
+    // A reverse lookup table for instantiations.
+    instance_for_offset: Vec<Vec<SignatureToken>>,
+    instantiations: HashMap<Vec<SignatureToken>, LocalsSignatureIndex>,
+    pub module: CompiledModule,
+}
+
+impl InstantiableModule {
+    pub fn new(module: CompiledModule) -> Self {
+        Self {
+            instantiations: module
+                .locals_signatures()
+                .iter()
+                .enumerate()
+                .map(|(index, sig)| (sig.0.clone(), LocalsSignatureIndex(index as TableIndex)))
+                .collect::<HashMap<_, _>>(),
+            instance_for_offset: module
+                .locals_signatures()
+                .iter()
+                .map(|loc_sig| loc_sig.0.clone())
+                .collect(),
+            module,
+        }
+    }
+
+    /// If the `instantiant` is not in the `instantiations` table, this sdds the instantiant to the
+    /// `instance_for_offset` for table, and adds the index to the reverse lookup table. Returns
+    /// the LocalsSignatureIndex for the `instantiant`.
+    pub fn add_instantiation(&mut self, instantiant: Vec<SignatureToken>) -> LocalsSignatureIndex {
+        match self.instantiations.get(&instantiant) {
+            Some(index) => *index,
+            None => {
+                let current_index =
+                    LocalsSignatureIndex(self.instance_for_offset.len() as TableIndex);
+                self.instantiations
+                    .insert(instantiant.clone(), current_index);
+                self.instance_for_offset.push(instantiant);
+                current_index
+            }
+        }
+    }
+
+    /// Returns the type instantiation at `index`. Errors if the instantiation does not exist.
+    pub fn instantiantiation_at(&self, index: LocalsSignatureIndex) -> &Vec<SignatureToken> {
+        match self.instance_for_offset.get(index.0 as usize) {
+            Some(vec) => vec,
+            None => {
+                panic!("Unable to get instantiation at offset: {:#?}", index);
+            }
+        }
+    }
+
+    /// Consumes self, and adds the instantiations that have been built up to the underlying
+    /// module, and returns the resultant compiled module.
+    pub fn instantiate(self) -> CompiledModuleMut {
+        let index = self.module.locals_signatures().len();
```

Comment:
just out of curiosity and to see what I am missing... why not use `instance_for_offset` directly here?

Reply:
This is actually a really good spot, and I think I'm being stupid here -- I think we can just use `instance_for_offset`. I'll take a look and see, but I think it should be exactly what is needed.